### PR TITLE
大规模训练：并行篇的 ASCII 图全部换成 SVG / Mermaid / 表

### PR DESCRIPTION
**Pull request by arganzheng**
## Summary

#74-5 上一轮的承诺：并行策略篇（`parallelism-strategies-which-state-to-shard`）剩余的 ASCII 图全部替换。

- 五张手绘 SVG：TP 列切 / 行切的矩阵切分；f / g 共轭在前向与反向的位置；Ring Attention 的环与 4 步 × 4 卡的 K/V 时间表；Ulysses all-to-all 的「表转置」；HSDP 的 2 × 8 网格。
- 注意力 + MLP 的配对流程换 Mermaid。
- SP 布局、因果 mask 负载均衡、Ring vs Ulysses 对照、EP 分组、rank 排布、405B 状态与通信账改成表；GPipe / 1F1B 的 ASCII 时间表删掉（已有 SVG）。
- 状态解剖篇的三段 MFU / HFU 计算改表。

## Test plan

- [x] `npm run check` 通过
- [x] `check-render.cjs` 两篇 PASS；五张新 SVG 截图看过
- [ ] `npm run review -- <PR#>`

Generated with [Devin](https://devin.ai)

### CODE REVIEW

File: _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md [hunk 1]
```diff
@@ -171,21 +171,13 @@ FSDP 是 ZeRO-3 在 PyTorch 里的原生实现，有两代。
...
 
-```text
- N_d = 16, N_s = 8（一个节点）, N_r = 2      placement = (Replicate(), Shard(0))
-                    ── 分片组：Shard(0)，AG + AG + RS ≈ 3N，NVLink ──→
- 副本 0（节点 0）   [ 0    1    2    3    4    5    6    7 ]   卡 i 持第 i 段
- 副本 1（节点 1）   [ 8    9   10   11   12   13   14   15 ]   卡 8+i 持同一段
-                     ↕    ↕    ↕    ↕    ↕    ↕    ↕    ↕
-                    复制组 {i, 8+i}：Replicate()，第 i 段梯度 AR ≈ 2N/N_s，IB
-```
+![HSDP 网格：16 张卡排成 2 行 8 列，每行是一个节点（副本），行内 8 张卡各持参数的一段（Shard(0)），横向是分片组的 3N 通信走 NVLink；每列的两张卡持同一段，竖向红箭头是复制组之间的梯度 all-reduce，走 IB](/img/in-post/parallelism-hsdp-grid.svg)
```
Comment (arganzheng) [inline _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md:174]: ASCII 网格换成 SVG（横向 NVLink 分片组、竖向 IB 复制组），显存 / 通信的对齐文本改表。

File: _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md [hunk 2]
```diff
@@ -195,25 +187,28 @@ ZeRO-3 的 $$3N$$ 通信全在 DP 组上；当 $$N_d$$ 跨越几十个节点时
 
 张量并行把一个线性层的权重矩阵切开。对 $$Y = XA$$（Megatron 的记法，$$X$$ 是 $$[\text{tokens}, h]$$ 的激活，$$A$$ 是 $$[h, h']$$ 的权重），有两种切法：
 
-```text
-列切（column-parallel）  A = [A_1 | A_2 | … | A_t]     每卡持 A_i（h × h'/t）
-                         输入 X 完整（每卡一份）；输出 Y_i = X A_i 是 Y 的第 i 列块
-                         → 输入不需通信，输出按列分布在各卡
+![上：列切——A 被竖着切成四条，每卡持一条，输入 X 完整、输出 Yᵢ = X Aᵢ 是 Y 的第 i 列块，不需通信；下：行切——A 被横着切成四条，输入 X 必须按列切成 Xᵢ，每卡算出一个部分和 XᵢAᵢ，四个部分和要 all-reduce 相加](/img/in-post/parallelism-tp-column-row-split.svg)
```
Comment (arganzheng) [inline _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md:190]: 两个 ASCII 块：矩阵切法换 SVG（上列切、下行切）+ 一张对照表；注意力 + MLP 的配对流程换 Mermaid。

File: _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md [hunk 3]
```diff
@@ -222,13 +217,7 @@ Megatron-LM（Shoeybi et al. 2019）的洞见是把两者**配对**：MLP 的第
...
 前向每层两次 all-reduce（注意力后一次、MLP 后一次），每次的载荷是一个完整的激活张量 $$s \cdot b \cdot h$$ 个元素。反向也是两次：行切层前向的 all-reduce 在反向是恒等（梯度直接分发），而列切层前向的恒等（输入广播）在反向变成 all-reduce（各卡对 $$X$$ 的梯度要求和）——两者互为共轭，所以反向的 all-reduce 出现在与前向不同的位置，但次数相同。用 Megatron 论文的记法（$$f$$：前向恒等、反向 all-reduce；$$g$$：前向 all-reduce、反向恒等），一个 MLP 块的前向与反向是：
 
-```text
- 前向  X ──f: 恒等──→ [列切 W1] ──→ GELU ──→ [行切 W2] ──g: all-reduce──→ Y
-          各卡各持完整 X                    各卡持部分和 Σ_i        各卡得完整 Y
-
- 反向  dX ←──f: all-reduce── [W1ᵢᵀ] ←── GELU' ←── [W2ᵢᵀ] ←──g: 恒等── dY
-          各卡对 X 的部分梯度求和               dY 完整，各卡直接取自己那块
-```
+![上：前向——X 完整经 f（恒等）进列切 W1ᵢ、GELU、行切 W2ᵢ，各卡得到部分和，经 g（all-reduce）得完整 Y；下：反向——dY 完整经 g（恒等）各卡取自己那块，经 W2ᵢᵀ、GELU′、W1ᵢᵀ 得到对 X 的部分梯度，经 f（all-reduce）求和得 dX](/img/in-post/parallelism-tp-conjugate-fg.svg)
```
Comment (arganzheng) [inline _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md:220]: f / g 前向反向的 ASCII 换 SVG，把「前向 all-reduce 在块尾、反向在块头」画出来。

File: _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md [hunk 4]
```diff
@@ -248,11 +237,10 @@ TP 切了层**内**的激活（Q/K/V、FFN 中间态按头 / 按列分布），
 
 **序列并行**（Megatron 意义下的 SP，注意与后面的 CP 区分）把这部分沿序列维切成 $$N_t$$ 份：LayerNorm、dropout、残差加法都是逐 token 的，切开序列不影响结果。切开后进入列切线性层前要把序列拼回来（**all-gather**），行切线性层的输出原本要 all-reduce、现在改成 **reduce-scatter**（归约的同时按序列切开），正好落回序列并行的布局：
 
-```text
-  无 SP：   LN ──── X（完整）───→ [列切] … [行切] ─── all-reduce ───→ + 残差 ─── LN
-  有 SP：   LN ── X_i（s/t）── all-gather ──→ [列切] … [行切] ── reduce-scatter ──→ + 残差 ── LN
-            ↑ 每卡 1/t                                                         ↑ 每卡 1/t
-```
+| | 层边界（LN、dropout、残差） | 进列切层前 | 行切层之后 | 层边界 |
```
Comment (arganzheng) [inline _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md:240]: 无 SP / 有 SP 两行布局的 ASCII 改表。

File: _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md [hunk 5]
```diff
@@ -344,14 +320,7 @@ PyTorch 2.13.0 里 CP 是实验 API：`torch/distributed/tensor/experimental/_co
...
 
-```text
- p = 4, m = 4      时间 →
- stage 0   F0 F1 F2 F3 ·  ·  ·  ·  ·  ·  B3 B2 B1 B0
- stage 1   ·  F0 F1 F2 F3 ·  ·  ·  ·  B3 B2 B1 B0 ·
- stage 2   ·  ·  F0 F1 F2 F3 ·  ·  B3 B2 B1 B0 ·  ·
- stage 3   ·  ·  ·  F0 F1 F2 F3 B3 B2 B1 B0 ·  ·  ·
-                    ↑ 前向填充 p-1 格          ↑ 反向排空 p-1 格
-```
+下图上半是 $$p = 4$$、$$m = 4$$ 的时间表（横轴时间、每行一个 stage，F 是前向、B 是反向），两头各有 $$p - 1$$ 格空白——气泡。
```
Comment (arganzheng) [inline _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md:323]: ASCII 时间表删掉，指向已有的 GPipe vs 1F1B SVG。

File: _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md [hunk 6]
```diff
@@ -366,14 +335,7 @@ $$
...
 
-```text
- p = 4, m = 8
- stage 0   F0 F1 F2 F3 B0 F4 B1 F5 B2 F6 B3 F7 B4 ·  B5 ·  B6 ·  B7
- stage 1   ·  F0 F1 F2 B0 F3 B1 F4 B2 F5 B3 F6 B4 F7 B5 ·  B6 ·  B7 ·
- stage 2   ·  ·  F0 F1 B0 F2 B1 F3 B2 F4 B3 F5 B4 F6 B5 F7 B6 ·  B7 ·  ·
- stage 3   ·  ·  ·  F0 B0 F1 B1 F2 B2 F3 B3 F4 B4 F5 B5 F6 B6 F7 B7 ·  ·  ·
-           ←warm-up→←────────────── 稳态：1F1B ──────────────→←cool-down→
-```
+下图下半是 $$p = 4$$、$$m = 8$$ 的 1F1B 时间表：
```
Comment (arganzheng) [inline _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md:338]: 同上。

File: _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md [hunk 7]
```diff
@@ -389,11 +351,12 @@ $$
 \frac{T_{\text{bubble}}}{T_{\text{ideal}}} = \frac{(p-1)(t_f + t_b)/v}{m(t_f + t_b)} = \frac{p-1}{v\,m}
 $$
 
-```text
- p = 2 卡, v = 2 → 4 个 stage：卡 0 持 stage {0, 2}，卡 1 持 stage {1, 3}
- 卡 0   F0⁰ F1⁰ F0² F1² B1² B0² B1⁰ B0⁰ …      上标为 stage 编号；同一 micro-batch 在卡 0 上进出两次
- 卡 1   ·   F0¹ F1¹ F0³ F1³ B1³ B0³ B1¹ B0¹ …
-```
+$$p = 2$$ 张卡、$$v = 2$$，共 4 个 stage，卡 0 持 stage {0, 2}、卡 1 持 stage {1, 3}（上标是 stage 编号，同一个 micro-batch 在卡 0 上进出两次）：
```
Comment (arganzheng) [inline _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md:354]: 两卡四 stage 的 ASCII 时间表改表。

File: _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md [hunk 8]
```diff
@@ -274,12 +262,7 @@ CP 与 SP 的区别：SP 只切层边界处那些逐 token 的算子，注意力
 
 Ring Attention（Liu et al. 2023）让每张卡固定持有自己那块 Q，把 K/V 块沿环传递：第 $$j$$ 步用来自第 $$(i-j) \bmod N_c$$ 张卡的 K/V 块算一个局部注意力，同时把手上的 K/V 块发给下一张卡、接收上一张卡的。$$N_c - 1$$ 步后每块 Q 看过了全部 K/V；局部结果用 online-softmax 的方式合并（与 FlashAttention 分块的合并方式相同）。
 
-```text
-  卡 0: Q_0  K/V_0 → K/V_3 → K/V_2 → K/V_1        每步：算 attn(Q_0, K/V_j)，同时收发下一块
-  卡 1: Q_1  K/V_1 → K/V_0 → K/V_3 → K/V_2
-  卡 2: Q_2  K/V_2 → K/V_1 → K/V_0 → K/V_3
-  卡 3: Q_3  K/V_3 → K/V_2 → K/V_1 → K/V_0
-```
+![左：四张卡围成一个环，每张卡固定持有自己的 Q 块，K/V 块沿环顺时针 send/recv；右：4 步 × 4 卡的表格，每一格是该卡该步用的 K/V 块——卡 0 依次用 K/V₀、K/V₃、K/V₂、K/V₁，每一列四张卡用的块各不相同](/img/in-post/parallelism-ring-attention.svg)
```
Comment (arganzheng) [inline _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md:265]: 环 + 4 步时间表换 SVG；因果负载均衡的对齐文本改表。

File: _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md [hunk 9]
```diff
@@ -289,23 +272,18 @@ $$
...
 ### 3. Ulysses：按头 all-to-all
 
 DeepSpeed-Ulysses（Jacobs et al. 2023）换一个思路：注意力对**头**是独立的。每张卡持有 $$s/N_c$$ 个 token 的全部头，算 Q/K/V 投影后做一次 **all-to-all**，变成持有全部 $$s$$ 个 token 的 $$a/N_c$$ 个头——这时每张卡可以对自己的头做完整的、不需要任何通信的注意力；算完再 all-to-all 回到按序列切的布局，进输出投影。
 
-```text
-      按序列切 [s/c, b, a·d]  ── all-to-all(Q)、(K)、(V) ──→  按头切 [s, b, a/c·d]  ── 局部完整注意力
-                                                                       │
-      按序列切 [s/c, b, a·d]  ←──────── all-to-all(O) ────────────────┘
-```
+![左：按序列切的 4×4 表，每行一张卡持自己那段 token 的全部 4 个头（四种颜色）；右：按头切，每行一张卡持全部 token 的一个头（一种颜色）；中间两条红箭头是 Q/K/V 的 all-to-all 与 O 的 all-to-all——把表转置](/img/in-post/parallelism-ulysses-all-to-all.svg)
```
Comment (arganzheng) [inline _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md:286]: all-to-all 的 ASCII 换 SVG（按序列切 → 按头切的表转置）。

File: _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md [hunk 10]
```diff
@@ -317,16 +295,14 @@ $$
...
-                      Ring Attention                       Ulysses
-──────────────────    ────────────────────────────────     ──────────────────────────────────
-切的状态              Q/K/V/激活 沿 s 切 1/N_c               同，但注意力内部临时按头切
-通信形态              send/recv 环，N_c - 1 步                all-to-all，每层 4 + 4 次
-每卡每层通信量        ≈ 12 s b h_kv（与 N_c 无关）              ≈ 16 s b h / N_c
-能否重叠              是（与分块注意力计算流水）               否（关键路径）
-N_c 上限              无                                     ≤ 头数（GQA：≤ K/V 头数）
-与 TP 组合            K/V 块再除以 N_t                         头数再除以 N_t，上限更紧
-```
+| | Ring Attention | Ulysses |
```
Comment (arganzheng) [inline _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md:298]: 对照的对齐文本改表。

File: _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md [hunk 11]
```diff
@@ -536,37 +494,36 @@ Llama 3 论文（Dubey et al. 2024）给出的 405B 训练配置有三档（Tabl
...
-```text
-每卡参数份额       N_local = 405B / (8 × 16) = 3.16B
-bf16 参数          2 × 3.16B = 6.3 GB
-bf16 梯度          6.3 GB
-优化器 (fp32 × 3)  12 × 3.16B / N_d = 38 GB / 128 = 0.3 GB           ← ZeRO-1 把 38 GB 切成 0.3 GB
-常驻合计           ≈ 12.9 GB                                           （二进制单位 12.06 GiB）
-每层激活（每卡）    34 × 8192 × 16384 / 8 = 570 MB（FlashAttention，SP 已切 1/N_t，b = 1）
-每 stage 层数      126 / 16 ≈ 8 层；1F1B 下 stage 0 在途 ≤ 16 个 micro-batch → 峰值 8 × 16 × 570 MB ≈ 73 GB
-```
+| 项 | 计算 | 结果 |
```
Comment (arganzheng) [inline _posts/2026-07-17-parallelism-strategies-which-state-to-shard.md:497]: 状态放置与两档通信量的对齐文本改成三张表（计算列保留原式）。

File: _posts/2026-07-15-training-state-anatomy-memory-and-mfu.md [hunk 12]
```diff
@@ -509,27 +509,27 @@ $$
...
 70B、1024 张 H100、global batch 4M token，实测 step 时间 4.5 s：
 
-```text
-tokens/s                = 4,194,304 / 4.5                    =  932 K token/s
-模型 FLOP/token         = 4.49 × 10¹¹（第六章精确值）
-模型 FLOP/s（全集群）    = 932 × 10³ × 4.49 × 10¹¹            =  4.19 × 10¹⁷
-峰值 FLOP/s（全集群）    = 1024 × 9.89 × 10¹⁴                  =  1.013 × 10¹⁸
-MFU                     = 4.19 × 10¹⁷ / 1.013 × 10¹⁸         =  41.3%
-```
+| 量 | 计算 | 结果 |
```
Comment (arganzheng) [inline _posts/2026-07-15-training-state-anatomy-memory-and-mfu.md:512]: MFU / HFU 三段计算的对齐文本改表（上一轮判为「计算，可留 text」，本轮与系列其他篇统一）。

**Review by arganzheng — COMMENTED** [review]
修改理由（每节一条，由 tools/review.py --notes 发出）
